### ISSUE TITLE
Suitability inquiry: patch generation?

### ISSUE THREAD

**Issue by rulatir** [JBlond/php-diff#120]
Given file1, file2, file3... on disk and $modifiedContents1, $modifiedContents2, $modifiedContents3... strings (existing ONLY in memory and without ANY file representation), how difficult would it be to use this library to render a standard patch file that will apply the respective modifications to file1, file2, file3 when fed to the standard `patch` utility?

**Comment by rulatir** [comment]
I tried to concatenate the output of multiple diffs with appropriate headers prepended, but I am facing a difficult issue. The patch format seems to require special handling _if and only if_ the following conditions simultaneously hold:

 1. the last line of the file ends up emitted as part of unmodified context of the last change, and
 2. there is no newline at the end of file

This condition is non-trivial to detect, mainly because of the subcondition 1. Some support from the library would be welcome.

**Comment by JBlond** [comment]
I wonder since the text Unified output is the same as `diff -u a.txt b.txt`. That output can be used to "feed" patch.

**Comment by JBlond** [comment]
@rulatir It would be nice to have some example content to look at.

**Comment by jfcherng** [comment]
> I wonder since the text Unified output is the same as `diff -u a.txt b.txt`. That output can be used to "feed" patch.

Not exactly. You would have to add support for "No newline at end of file" situation, which I did in my lib.

**Comment by JBlond** [comment]
I added the function for that into https://github.com/JBlond/php-diff/tree/php-diff-120
It still must be added to the renderers. PRs are welcomed.

**Comment by DigiLive** [comment]
@JBlond 
After a busy period, I have some spare time now to help you out, but I need a heads-up.
Please send me an email with whats supposed to happen.

**Comment by JBlond** [comment]
@DigiLive I did, and I hope that it is clear now. If not, ask rulatir for more details.

**Comment by DigiLive** [comment]
Nope... it's still not clear to me what's expected to happen.
Maybe @rulatir or @jfcherng can elaborate.
I need more details than already given.

**Comment by jfcherng** [comment]
If https://github.com/JBlond/php-diff/issues/120#issuecomment-1336193044 is not clear enough, then I think you have to diff them with `diff -u` by yourself.

Here's a [sample.zip](https://github.com/JBlond/php-diff/files/10447620/sample.zip), just run `diff -u old.txt new.txt` and see the output.

**Comment by DigiLive** [comment]
I'm currently looking into this issue.
However, my knowledge is about PHP and not diffutils. ;)

It seems like GNU diffutils:
1. Appends line `\ No newline at end of file` to the output of a file's last line, if this line doesn't contain EOL characters.
2. Drops the last line for comparison if it doesn't contain any characters, including EOL.

**Comment by JBlond** [comment]
I'm a bit puzzled by the definition. https://github.com/Distrotech/diffutils/blob/9e70e1ce7aaeff0f9c428d1abc9821589ea054f1/doc/diffutils.texi#L1718-1748

**Comment by JBlond** [comment]
@DigiLive After reading a gazillion of implementations I can confirm that you are right with your conclusions..

**Comment by JBlond** [comment]
The code for this is at https://github.com/Distrotech/diffutils/blob/9e70e1ce7aaeff0f9c428d1abc9821589ea054f1/src/util.c#L663 and 
https://github.com/Distrotech/diffutils/blob/9e70e1ce7aaeff0f9c428d1abc9821589ea054f1/src/diff3.c#L1416
